src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs

```diff
@@ -696,7 +696,26 @@ private BoundBlock RewriteBlock(BoundBlock node, ArrayBuilder<BoundExpression> p
 
             if (prologue.Count > 0)
             {
-                newStatements.Add(new BoundSequencePoint(null, null) { WasCompilerGenerated = true });
+                // Add hidden sequence point, prologue -- closure initialization -- doesn't map to source.
+
+                // However, when the closure is used inside of conditional blocks like 'if (b) {...}',
+                // we need to put a regular seqeunce point to closure initialization 
+                // associated with the closure scope start ('{'), so IP can move to '{' without skipping initialization.
+
+                var initialization = prologue.First();
+                var block = initialization.Syntax as BlockSyntax;
+                if (block != null && (block.Parent.IsKind(SyntaxKind.IfStatement) || block.Parent.IsKind(SyntaxKind.ElseClause)))
```

Comment:
This condition seems very tailored to the particular test case, and I suspect many similar situations might trigger the bug, for example

```cs
    for (Action a = null; expr; )
    {
        a = (Action)delegate { a = null; };
        a();
    }
```

The lambda rewriter should not be conditioned on the shape of the original source text, but rather on the shape of the code in its current lowered form.